e2e test design document (#69)
cybozu-go/moco: e2e/README.md

# End-to-end test suites

## Strategy

We try to maximize coverage of unit tests even if it needs "real" MySQL instances.  The tests using real MySQL instances can be divided into the unit tests and the e2e tests as follows:

> - test by unit test

Reviewer: I'd suggest "small test" rather than unit test to denote the tests precisely.

  - initialization of MySQL instance
  - status retrieval
  - cloning
  - metrics collection
  - log file management (rotate)
- test by e2e test
  - failover/switchover
  - replication
  - backup/restore
  - point-in-time recovery
  - upgrade
  - blackout recovery
  - failure injection

The unit tests are written in each `_test.go` file.

The remains are:
- YAML manifests
- `main` functions of each programs

## Analysis

### Feature functions

The following feature functions should be examined by checking the status of `MySQLCluster` custom resource and MySQL instances.
- failover/switchover
- replication
- backup/restore
- point-in-time recovery
- upgrade
- blackout recovery
- failure injection

### YAML manifests

> `MySQLCluster` custom resource accepts some flexible settings such as the templates of Pod, Volume, and Service.  So, we should examine the settings are correctly applied to the resources made by MOCO.

Reviewer: Such reconciliation results should be tests in small tests.
Does this document cover small tests in addition to e2e tests?

Reviewer 2: No. We will write about e2e tests.

# End-to-end test suites

## Strategy

> We try to maximize coverage of unit tests even if it needs "real" MySQL instances.

Reviewer: ```suggestion
We try to maximize coverage of small tests even if it needs "real" MySQL instances.
```

> The tests using real MySQL instances can be divided into the small tests and the e2e tests as follows:

Reviewer: ```suggestion
Specifically, small tests should cover the following areas.
```

- by small test
  - initialization of MySQL instance
  - status retrieval
  - cloning
  - metrics collection
  - log file management (rotate)

> - by e2e test

Reviewer: ```suggestion
End-to-end (e2e) tests should include but not limited to:
```

  - failover
  - switchover
  - replication
  - backup/restore
  - point-in-time recovery
  - upgrade
  - blackout recovery
  - failure injection

> The small tests are written in each `_test.go` file.

Reviewer: ```suggestion
The small tests exist in each package directory.
```

> The remains are:

Reviewer: Why do these remain?  I guess these should be covered in e2e.

- YAML manifests
- `main` functions of each programs

## Analysis

> ### Feature functions

Reviewer: I'd propose "Scenarios" instead.
